### PR TITLE
new-exec

### LINKED ISSUE #3638: [nix-local-build] new-run, new-test, new-bench, new-exec
**Summary by @ezyang.** These should be easy to do, it's a copy-paste job of the existing run/test/bench code on top of invoking the new-build logic to build things.

---

These are very important quality-of-life features that make it impractical to use `new-build` for some of our larger projects. In general, we'd be happy with straightforward ports of the legacy commands, but a couple specific comments arose.
#### Avoiding unnecessary work

This is a lot more important when working with metapackages, since we don't want to have to build every test suite in the metapackage in order to build one particular test. The way the `[COMPONENTS]` argument currently works for `new-build` should be fine for this.
#### Easier argument passing for test and bench

`cabal run [FLAGS] [EXECUTABLE] [-- EXECUTABLE_FLAGS]` makes it easy to pass arguments to the executable by just putting them after the `--`. It would be nice to have the same for `test` and `bench`, rather than having to use `--test-options` or `--benchmark-options` flags.

### PR DESCRIPTION
**Pull request by fgaz**
Merged master with dmwit/master and finished new-exec.

* [x] Any changes that could be relevant to users have been recorded in the changelog.
* [x] The documentation has been updated, if necessary.
* [x] Basic tests.

I had some trouble with git. There are two mege commits where I should have rebased instead, and now it's kinda difficult to fix (see those new-bench commits? they are from the merge.), but I suppose I'll have to do it before merging in master?

Also there's the question of how to rewrite the ghc program name. As a temporary solution I just check if it's ``elem` ["ghc", "ghcjs", "ghci"]` and use it as is.
If there is no objection I'll change it to use the selected (`-w`) ghc version.

Fixes #3638 (finally!)

### CODE REVIEW

File: cabal-install/Distribution/Client/CmdExec.hs [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    [] -> die' verbosity "Please specify an executable to run"
+    exe:args -> do
+      (program, _) <- requireProgram verbosity (simpleProgram exe) programDb
+      let argOverrides =
+            argsEquivalentOfGhcEnvironmentFile
+              (distDirLayout baseCtx)
+              (elaboratedPlanOriginal buildCtx)
+              buildStatus
+          programIsCompiler =
+            programId program `elem` ["ghc", "ghcjs", "ghci"]
```
Comment (fgaz) [inline cabal-install/Distribution/Client/CmdExec.hs]: How do we handle the compiler name rewriting?

**Comment by 23Skidoo** [comment]
I'm on vacation in Norway right now, but I'll review this when I get back next week.

**Comment by cocreature** [comment]
FWIW, as a user I would like `cabal new-exec` to *not* rewrite commands and only set environment variables. Rewriting commands is bound to cause confusion and will always be really fragile, e.g., `cabal new-exec -- ghc` might work but `cabal new-exec -- time ghc` won’t. I can see a case for a way to execute `ghc` regardless of how the executable is called but I would prefer for that to be a separate `cabal new-ghc` command.

**Comment by fgaz** [comment]
@cocreature the rewriting only happens as a fallback when using older versions of ghc that don't support .ghc.environment files, and eventually (well, after the 5 or so years of support...) will disappear. Other compilers will probably need rewriting though (I'm changing that code to be more general).

I agree that the naming is unfortunate, and I actually would prefer a `new-ghc` too, but `new-exec` is meant to be kinda compatible with old `exec`, and `cabal -w "some-ghc" new-exec -- "some-ghc"` will be completely equivalent to the hypotetical `cabal -w "some-ghc" new-ghc`.

Of course if there is enough support we can always split the command, it's only a few changes.

**Comment by cocreature** [comment]
@fgaz There will be an option to disable `.ghc.environment.*` files https://github.com/haskell/cabal/issues/4542 (which I am definitely going to use) so I don’t think you can rely on this ever going away.

**Comment by fgaz** [comment]
@cocreature that flag only applies to the .ghc.env file that every new- command generates in the project root. new-exec uses a separate and temporary .ghc.env file which is always generated regardless of that flag.

File: cabal-install/Distribution/Client/ProjectPlanOutput.hs [hunk 2]
```diff
@@ -701,9 +743,44 @@ renderGhcEnviromentFile projectRootDir elaboratedInstallPlan
...
+  :: Compiler
+  -> DistDirLayout
+  -> ElaboratedInstallPlan
+  -> PostBuildProjectStatus
+  -> [String]
+argsEquivalentOfGhcEnvironmentFile compiler =
+  case compilerId compiler
+  of CompilerId GHC   _ -> argsEquivalentOfGhcEnvironmentFileGhc
+     CompilerId GHCJS _ -> argsEquivalentOfGhcEnvironmentFileGhc
+     CompilerId _     _ -> error "Only GHC and GHCJS are supported"
```
Comment (fgaz) [inline cabal-install/Distribution/Client/ProjectPlanOutput.hs]: Is `error` appropriate? `ImplInfo` does it like this too.

File: cabal-install/Distribution/Client/CmdExec.hs [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+                  installFlags haddockFlags
+    withOverrides env args program = program
+      { programOverrideEnv = programOverrideEnv program ++ env
+      , programDefaultArgs = programDefaultArgs program ++ args}
+
+matchCompilerPath :: ElaboratedSharedConfig -> ConfiguredProgram -> Bool
+matchCompilerPath elaboratedShared program =
+  programPath program
+  `elem`
+  (programPath <$> configuredCompilers)
```
Comment (fgaz) [inline cabal-install/Distribution/Client/CmdExec.hs]: Should i canonicalize those paths?

**Comment by 23Skidoo** [comment]
OSX failure seems to be spurious.

File: cabal-install/Distribution/Client/CmdExec.hs [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+import Prelude ()
+import Distribution.Client.Compat.Prelude
+
+import Data.Set (Set)
+import qualified Data.Set as S
+import qualified Data.Map as M
+
+execCommand :: CommandUI (ConfigFlags, ConfigExFlags, InstallFlags, HaddockFlags)
+execCommand = installCommand
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/CmdExec.hs]: This is a bit hacky. `installCommand` is not obviously related to `new-build` and will be going away. Also we don't actually want to pull in everything from there. So I'd be more explicit and use `commandOptions = commandOptions CmdBuild.buildCommand, ...`. I understand that `CmdBuild.buildCommand` is giving a bad example, but we should fix it too.

File: cabal-install/Distribution/Client/ProjectPlanOutput.hs (hunk 2 above)
Comment (23Skidoo) [reply cabal-install/Distribution/Client/ProjectPlanOutput.hs]: It's OK.

File: cabal-install/Distribution/Client/ProjectPlanOutput.hs [hunk 5]
```diff
@@ -701,9 +743,44 @@ renderGhcEnviromentFile projectRootDir elaboratedInstallPlan
...
+
+-- remove this when we drop support for non-.ghc.env ghc
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/ProjectPlanOutput.hs]: Forgotten TODO?

File: cabal-install/Distribution/Client/CmdExec.hs (hunk 3 above)
Comment (23Skidoo) [reply cabal-install/Distribution/Client/CmdExec.hs]: Not 100% sure, probably not. Aren't the configured progs paths already absolute?

Comment (fgaz) [reply cabal-install/Distribution/Client/CmdExec.hs]: No idea. They _seem_ absolute but maybe it's only the general case. I'll probably leave it like this and add a comment about it.

Comment (23Skidoo) [reply cabal-install/Distribution/Client/CmdExec.hs]: OK. I think they should be absolute after the configure progs step.

**Comment by fgaz** [comment]
All done.

...except that `git rerere` failed me (or I failed it) and now we have three more commits that could have been `fixup`d. Can I merge anyway?

**Comment by 23Skidoo** [comment]
There's a merge conflict.

**Comment by 23Skidoo** [comment]
Merged, thanks!

**Pull request merged**